### ISSUE TITLE
Ignore type checking imports

### ISSUE THREAD

**Issue by Dobatymo** [jendrikseipp/vulture#152]
Code like this is common when type annotations in comments are used in Python:

```
from typing import TYPE_CHECKING
if TYPE_CHECKING:
	from typing import Optional, Iterator
	from abc import xyz
```
These will usually show up as `unused import` as they are only referenced from comments.

The code should probably be parsed with a `TYPE_CHECKING=False` in mind.

**Comment by jendrikseipp** [comment]
Could you please link to a project that uses this type of code, ideally directly to the source file?

**Comment by Dobatymo** [comment]
Random large python project: https://github.com/RasaHQ/rasa_nlu/blob/master/rasa_nlu/components.py#L23-L25

**Comment by jendrikseipp** [comment]
Thanks! I think it would be good to add a whitelist module for the classes (and maybe other types) in the `typing` module (https://docs.python.org/3/library/typing.html). Pull requests welcome :-) Detecting custom types (like the ones you linked above) is probably out of the scope of Vulture.

**Comment by jendrikseipp** [comment]
We discussed this further in #153 and found that adding a whitelist for the typing module won't help after all:

The code linked above uses type comments. To identify that the classes are actually used in the type comments we would need to parse the comments, which I think is tedious. Here, a manual whitelist is preferable, I think. More modern type-annotated code needs Python 3.6. There you can annotate function parameters and variables. This is actually already supported by Vulture.

**Comment by dsuch** [comment]
Hello,

I realise that this ticket is closed but perhaps for a future discussion, I just want to leave a note.

First, thanks for creating Vulture, this is a very useful tool!

As to the gist of this ticket - yes, code which does exactly what @Dobatymo describes actually exists and I can attest to that.

Using type hints directly in function or method definitions is not 100% the same as placing them in comments.

For instance:

```python
from typing import List

def func(a:List[str]) -> None:
  pass
```

And:

```python
if 0:
  from typing import List

def func(a):
  # type: (List[str]) -> None
  pass
```

Or:

```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
  from typing import List

def func(a):
  # type: (List[str]) -> None
  pass
```

The first of these incurs the most runtime overhead, the second one the least and the last one is in the middle.

Naturally, as far as type checking goes, all of the achieve the same but I am just saying that there is software that pays attention to such details and, personally, I prefer to use comments to have as little overhead as possible.

This is just one of dozens and dozens of micro-optimisations on top of various macro ones. 

The [mypy](https://mypy.readthedocs.io/en/latest/) type checker already recognises the ```if TYPE_CHECKING``` idiom and does not report any extra imports as erroneous, for instance:

```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from typing import List, Dict # Note that Dict is not used anywhere

def func(a):
    # type: (List[str]) -> None
    pass
```

```bash
$ mypy --strict test1.py 
Success: no issues found in 1 source file
$
```

As things stand today, Vulture - similarly to pyflakes - requires something like:

```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from typing import List

    List = List

def func(a):
    # type: (List[str]) -> None
    pass
```

This works and I am accustomed to it because I have been using it for years with pyflakes but if I did not have to write it, things would be clearer because that is just one import of many, many in a larger codebase.

To be more precise, before I started to use mypy, I was actually doing "if 0:" for such type imports but mypy does not recognise it so I settled on "if TYPE_CHECKING:" ultimately.

I understand the reluctance to support arbitrary things like "if something_that_may_return_False():" or even "if 0:" but with mypy's using "if TYPE_CHECKING:", this will likely become a settled idiom in the Python world.

Whether comments need to be parsed, like you said, I am not quite sure. Perhaps you could just assume that anything covered by "if TYPE_CHECKING" is to be ignored and if anyone uses comments for types, they can just use mypy.

At the end of they day, if there are bigger projects that use Vulture and would look forward to this feature, they will be likely to use mypy, Vulture, coverage and other tools too. At least, this is what my projects do.

Thanks again!

**Comment by dsuch** [comment]
I re-read my message and it may appear that it is only about runtime performance. That too, but what I wanted to relayed was the general usefulness of supporting "if TYPE_CHECKING:" directly.

For instance, here is another case that I am working on at this very moment.

A C-based extension called "foo" has an object called "BarClass".

```python
from foo import BarClass

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from typing import List

    List = List

def func(a):
    # type: (List[str]) -> None
    BarClass(a)

func(['abc'])
```

(Let us skip for the moment the fact that BarClass does not return anything and assume it is called only for its side effects).

Vulture 1.5 will not find any issues here, which is as expected because, lexically speaking, all of the tokens are in use.

However, mypy will complain that it is not able to check what "BarClass" actually is or what its ```__init__``` method looks like.

This means that one has to resort to tricks such as here ..

```python
if TYPE_CHECKING:
    from typing import cast, List

    List = List

    class _MyBar:
        def __init__(self, data):
            # type: (str) -> None
            self.data = data
```

.. and then to tell mypy that the previously-unknown BarClass behaves like _MyBar for the purposes of types validation.

This, however, may trigger "unused class '_MyBar' (60% confidence)" in Vulture, depending on whether the type hints are in comments or not.

Such things do happen in practice and they truly belong in the "if TYPE_CHECKING:" domain because they are to do only with static checks.

In other words, I am just saying that this idiom could be potentially special-cased. I mean these particular tokens "if TYPE_CHECKING:" because even mypy does not recognise something like:

```python
import typing
if typing.TYPE_CHECKING:
  ...
```

Regards.

**Comment by jendrikseipp** [comment]
@dsuch thanks for the explanations and examples. I have a much clearer picture of the problem now. After reading about how other linters handled this problem (https://github.com/PyCQA/astroid/issues/317, https://github.com/PyCQA/pyflakes/issues/247), I think the best solution is to parse the type comments that the `ast` module provides, starting in Python 3.8 (https://docs.python.org/3/library/ast.html). As an intermediate solution for Python versions before 3.8, we could ignore all names defined under `if typing.TYPE_CHECKING`, as you suggest. If anybody is interested in working on any of these two things, please speak up.

**Comment by The-Compiler** [comment]
FWIW I think pylint/astroid, mypy and probably others use [typed_ast](https://github.com/python/typed_ast) on Python < 3.8. Not sure how different the API is to the builtin `ast` with 3.8+ though.

**Comment by dsuch** [comment]
Hello,

thanks @jendrikseipp for taking it into consideration.

Regarding typed_ast - I seem to recall a discussion on [Roundup](https://bugs.python.org) about the fact that one had to use in runtime the same Python version to check tokens (and presumably AST) that the code being checked was using.

E.g. it was not guaranteed that Python 3.7 could tokenise Python 3.6 programs - perhaps this is related to the same effort.

I am not sure, unfortunately, but judging by its README it should be a drop-in replacement as far as Vulture goes because Vulture, if I am not mistaken, does not transform the AST in any way, it is read-only, so to speak.

**Comment by jingw** [comment]
For what it's worth, I am not personally a fan of special casing TYPE_CHECKING and blanket ignoring things defined/imported under it. The reason is that unused type imports and unused mypy-only declarations are still dead code that can be cleaned up (conditioned on counting type comments as usages). My understanding is that after #220, we'll have the optimal behavior: used type imports get ignored, and unused ones get flagged.

**Comment by dsuch** [comment]
Hello @jingw - this is not only about type hints from comments, please check the example with "class _MyBar:". This is the kind of code that needs to be placed somewhere solely to satisfy static checkers.

Here is a recent example. A test case class may begin with:

```python

from unittest import TestCase

class MyTestCase(TestCase):
    maxDiff = 100000

   # ... tests go here
```

The "maxDiff" attribute is needed because sometimes self.assertEqual will compare longer strings. But, because maxDiff is used only by TestCase, i.e. the parent class, it appears to be an unused variable (with confidence of 60%).

The only way to get around is to do something like that:

```python

from unittest import TestCase

class MyTestCase(TestCase):
    maxDiff = 100000

   # ... tests go here

# For Vulture
MyTestCase.maxDiff
```

This sort of things happen all the time. I am looking at a codebase of approximately 40k Python files as we speak and my experience is that placing such constructs close to what they apply to is more maintainable then using an external config file with patterns to ignore.

**Comment by jingw** [comment]
Hi @dsuch, I think the `_MyBar` case would be handled by #220. If `_MyBar` doesn't appear anywhere in a type comment or ordinary python, even Mypy isn't going to know that it should be associated with `BarClass`. I'm not sure what your code looks like, but I'm aware of people sometimes writing code like this:
```
if TYPE_CHECKING:
    class _MyBar:
        # define stubs
    BarClass = _MyBar
else:
    from library import BarClass

a = ...  # type: BarClass
```
In this code, neither _MyBar nor BarClass is flagged as dead code after #220. (By the way, I think the mypy folks would recommend https://mypy.readthedocs.io/en/stable/stubs.html for this use case.) Am I understanding correctly?

Regarding TestCase.maxDiff, large codebases, and excluding issues locally, I agree, but aren't those issues separate from the typing discussion in this thread? (i.e. you run into those issues regardless of whether your code is typed or whether TYPE_CHECKING is special cased.)

I'm also dealing with a codebase with tens of thousands of files, and my stance has been the following:
- Oneoff issues get suppressed with a local comment explaining what's up with the dead code. I would like to see official support for the `noqa: V123` comments currently hidden from the README in https://github.com/jendrikseipp/vulture/blame/master/README.md#L74. This would put vulture in line with flake8's `noqa`, mypy's `type: ignore`, and mypy's inline configuration. I agree with you about not wanting an external config file far away from the dead code. I want the documentation close to the code, and I disagree with the vulture README calling this "noise." Having a comment saying "this method is magically accessed via getattr()" is useful to a reader.
- However, systemic issues get suppressed centrally, since we don't want every developer / test case to need to explain why maxDiff is not dead code.

Happy to talk about this more on another issue!

**Comment by dsuch** [comment]
@jingw You are right on all accounts and, as I am not a core developer here and this project is looking very good already, I will be happy with any decision that you all will reach :-)

**Comment by jendrikseipp** [comment]
I agree, there are two distinct topics in this discussion. #220 should deal with the type comments. Regarding the noqa comments, I think you make a good point @jingw. It hadn't occured to me that having such a comment actually adds some information to readers of the code. I think it makes sense to "reopen" the discussion on noqa comments and how best to support them (for examples, speaking names vs. error numbers). Also, your stance on dealing with false positives makes a lot of sense to me.

**Comment by jendrikseipp** [comment]
#220 is now merged, thanks @jingw ! Note that for full typing support we would also need to parse forward references as described in #216.
